[Doc] Add Data Structures and Elements (#2070)
open-mmlab/mmsegmentation: docs/en/advanced_guides/structures.md

it has implemented basic functions of `Create`, `Read`, `Update`, `Delete`, supported data transferring among different types of devices
and tensor-like or dictionary-like operations such as `.cpu()`, `.cuda()`, `.get()` and `.detach()`.
More details can be found [here](https://github.com/open-mmlab/mmengine/blob/main/docs/en/advanced_tutorials/data_element.md).

MMSegmentation also follows this interface protocol and defines `SegDataSample` which is used to encapsulate the data of semantic segmentation task.

## Semantic Segmentation Data SegDataSample

[SegDataSample](mmseg.structures.SegDataSample) includes three main fields `gt_sem_seg`, `pred_sem_seg` and `seg_logits`, which are used to store the annotation information and prediction results respectively.

| Field          | Type                      | Description                     |
| -------------- | ------------------------- | ------------------------------- |
| gt_sem_seg     | [`PixelData`](#pixeldata) | Annotation information.         |
| pred_instances | [`PixelData`](#pixeldata) | The predicted result.           |

> | seg_logits     | [`PixelData`](#pixeldata) | The logits of predicted result. |

Reviewer: ```suggestion
| seg_logits     | [`PixelData`](#pixeldata) | The model output before normalized |
```

> | Field          | Type                      | Description                     |
> | -------------- | ------------------------- | ------------------------------- |
> | gt_sem_seg     | [`PixelData`](#pixeldata) | Annotation information.         |
> | pred_instances | [`PixelData`](#pixeldata) | The predicted result.           |
> | seg_logits     | [`PixelData`](#pixeldata) | The raw (non-normalized) predicted result. |

Reviewer: ```suggestion
| Field          | Type                      | Description                                |
| -------------- | ------------------------- | ------------------------------------------ |
| gt_sem_seg     | [`PixelData`](#pixeldata) | Annotation information.                    |
| pred_instances | [`PixelData`](#pixeldata) | The predicted result.                      |
| seg_logits     | [`PixelData`](#pixeldata) | The raw (non-normalized) predicted result. |
```

# Structures

To unify input and output interfaces  between different models and modules, OpenMMLab 2.0 MMEngine defines an abstract data structure,
it has implemented basic functions of `Create`, `Read`, `Update`, `Delete`, supported data transferring among different types of devices
and tensor-like or dictionary-like operations such as `.cpu()`, `.cuda()`, `.get()` and `.detach()`.
More details can be found [here](https://github.com/open-mmlab/mmengine/blob/main/docs/en/advanced_tutorials/data_element.md).

> Similarly, MMSegmentation also follows this data structure requirement. SegDataSample is used to encapsulate the data needed

Reviewer: ```suggestion
MMSegmentation also follows this interface protocol and defines `SegDataSample` which is used for encapsulate the data of semantic segmentation task
```

> # Data Structures and Elements

Reviewer: Why is there 'Elements' in title?

> During the training/testing process of a model, there is often a large amount of data to be passed between modules, and the data required by different tasks or algorithms is usually different. For example, in MMSegmentation, in training it needs image and its meta information but in inference it also needs rescaling information of input images. This makes the interfaces of different tasks or models may be inconsistent:

Reviewer: in training -> during training

what is rescaling information?

```python
# Training
for img, img_metas in dataloader:
  seg_logit = encode_decode(img, img_meta)

# Inference on whole image
for img, img_metas, rescale in dataloader:
  seg_logit = whole_inference(img, img_meta, rescale)

# Inference on sliding window
for img, img_metas, rescale in dataloader:
  seg_logit = slide_inference(img, img_meta, rescale)
```

From the above code examples, we can see that without encapsulation, the different data required by different tasks and algorithms lead to inconsistent interfaces between their modules, which seriously affects the extensibility and reusability of the library.

> Therefore, in order to solve the above problem, we use [DataElement tutorial](https://github.com/open-mmlab/mmengine/blob/main/docs/zh_cn/advanced_tutorials/data_element.md) from MMEngine to encapsulate the data required for each task into `data_sample`.

Reviewer: please change to English documentation
